### ISSUE TITLE
Max diagnostics per-line example crashes when removing lines above errors

### ISSUE THREAD

**Issue by casr** [neovim/neovim#29958]
### Problem

Under the documentation for `:h diagnostic-handlers-example` there is a handy example for reducing the number diagnostics shown per-line that looks like:

```lua
    -- Create a custom namespace. This will aggregate signs from all other
    -- namespaces and only show the one with the highest severity on a
    -- given line
    local ns = vim.api.nvim_create_namespace("my_namespace")

    -- Get a reference to the original signs handler
    local orig_signs_handler = vim.diagnostic.handlers.signs

    -- Override the built-in signs handler
    vim.diagnostic.handlers.signs = {
      show = function(_, bufnr, _, opts)
        -- Get all diagnostics from the whole buffer rather than just the
        -- diagnostics passed to the handler
        local diagnostics = vim.diagnostic.get(bufnr)

        -- Find the "worst" diagnostic per line
        local max_severity_per_line = {}
        for _, d in pairs(diagnostics) do
          local m = max_severity_per_line[d.lnum]
          if not m or d.severity < m.severity then
            max_severity_per_line[d.lnum] = d
          end
        end

        -- Pass the filtered diagnostics (with our custom namespace) to
        -- the original handler
        local filtered_diagnostics = vim.tbl_values(max_severity_per_line)
        orig_signs_handler.show(ns, bufnr, filtered_diagnostics, opts)
      end,
      hide = function(_, bufnr)
        orig_signs_handler.hide(ns, bufnr)
      end,
    }
```

When at least two lines are removed, and there is a line following them that has a severity of at least error, it results in:

```
Error executing vim.schedule lua callback: /opt/local/share/nvim/runtime/lua/vim/diagnostic.lua:1328: Invalid 'line': out of range
stack traceback:
        [C]: in function 'nvim_buf_set_extmark'
        /opt/local/share/nvim/runtime/lua/vim/diagnostic.lua:1328: in function 'show'
        ~/.config/nvim_severity_test/init.lua:28: in function 'show'
        ~/.config/nvim_severity_test/plugin/diagnostics.lua:43: in function 'show'
        /opt/local/share/nvim/runtime/lua/vim/diagnostic.lua:1684: in function 'show'
        /opt/local/share/nvim/runtime/lua/vim/diagnostic.lua:1031: in function 'set'
        /opt/local/share/nvim/runtime/lua/vim/lsp/diagnostic.lua:260: in function 'handle_diagnostics'
        /opt/local/share/nvim/runtime/lua/vim/lsp/diagnostic.lua:293: in function 'handler'
        /opt/local/share/nvim/runtime/lua/vim/lsp/client.lua:1005: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

### Steps to reproduce

1. create `~/.config/nvim_severity_test/init.lua`:
    ```lua
    vim.api.nvim_create_autocmd('FileType', {
      pattern = "lua",
      callback = function()
        vim.lsp.start({
          name = "lua_ls",
          cmd = { "lua-language-server" },
          root_dir = vim.fn.getcwd(),
        })
      end,
    })
    
    local ns = vim.api.nvim_create_namespace("my_namespace")
    local orig_signs_handler = vim.diagnostic.handlers.signs
    vim.diagnostic.handlers.signs = {
      show = function(_, bufnr, _, opts)
        local diagnostics = vim.diagnostic.get(bufnr)
        local max_severity_per_line = {}
        for _, d in pairs(diagnostics) do
          local m = max_severity_per_line[d.lnum]
          if not m or d.severity < m.severity then
            max_severity_per_line[d.lnum] = d
          end
        end
        local filtered_diagnostics = vim.tbl_values(max_severity_per_line)
        orig_signs_handler.show(ns, bufnr, filtered_diagnostics, opts)
      end,
      hide = function(_, bufnr)
        orig_signs_handler.hide(ns, bufnr)
      end,
    }
    ```
2. create `test_case.lua`
    ```lua
    -- 1. line to delete
    -- 2. line to delete
    syntax error
    ```
3. open Neovim with `NVIM_APPNAME=nvim_severity_test nvim test_case.lua`
4. wait a moment for the LSP server to report diagnostics
5. delete the comment block in one go with `dgc`

### Expected behavior

No crash.

### Neovim version (nvim -v)

NVIM v0.10.0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

macOS 14.5

### Terminal name/version

Terminal.app/tmux 3.4

### $TERM environment variable

tmux-256color

### Installation

MacPorts

**Comment by clason** [comment]
Please test the latest version (0.10.1), and ideally the latest `master` (we have nightlies on our releases page for this very purpose).

**Comment by casr** [comment]
Using the releases page's builds, the issue presents in v0.10.1 but seems to be fixed in the nightlies.

Tested version was:
```
NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1713484068
Run "nvim -V1 -v" for more info
```
and
```
NVIM v0.11.0-dev-555+g48e4589ea
Build type: RelWithDebInfo
LuaJIT 2.1.1720049189
Run "nvim -V1 -v" for more info
```

Unless this is something worth back porting to the v0.10.x, feel free to close this out.

**Comment by clason** [comment]
Backported; when that is merged, you can try building the `release-0.10` branch from source and see if that fixed it.

**Comment by casr** [comment]
Thank you!
